fix: use SignJwt instead of GenerateAccessToken for Vault GCP Workload Identity auth

SamuelMolling opened the pull request:
## Description
This PR fixes the Vault GCP IAM authentication when using Workload Identity by generating a properly signed JWT instead of using an OAuth2 access token.

## Problem
When using Vault's GCP auth method with Workload Identity, the External Secrets Operator was passing an OAuth2 access token instead of a signed JWT, causing the error:
```
unable to parse signed JWT: go-jose/go-jose: compact JWS format must have three parts
```

## Solution
- Added `SignJwt` method to `IamClient` interface
- Created `SignedJWTForVault` method to generate JWT with correct claims (sub, aud, exp)
- Added `GenerateSignedJWTForVault` public function in auth.go  
- Modified Vault `auth_gcp.go` to use signed JWT for Workload Identity
- Added `loginWithSignedJWT` method to authenticate directly with Vault API

## Changes
- `providers/v1/gcp/secretmanager/auth.go` - Added public function to generate signed JWT for Vault
- `providers/v1/gcp/secretmanager/workload_identity.go` - Added SignJwt method and JWT generation logic
- `providers/v1/vault/auth_gcp.go` - Modified to use signed JWT and added login method
- `providers/v1/vault/client.go` - Added gcpSignedJWT field to store the JWT

## Testing
Tested successfully with:
- GKE cluster with Workload Identity enabled
- Vault configured with GCP IAM auth method
- ClusterSecretStore using VaultGCPAuth with WorkloadIdentity configuration

## Related Issues
Fixes the JWT parsing error when using Vault GCP auth with Workload Identity.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
This PR fixes Vault GCP IAM authentication when using Workload Identity by generating and using a properly signed JWT (via SignJwt) instead of an OAuth2 access token, resolving the error "unable to parse signed JWT: go-jose/go-jose: compact JWS format must have three parts."

Key changes
- Shared auth: Extracted GCP auth and Workload Identity logic into runtime/gcp/auth (NewTokenSource, GenerateSignedJWTForVault, workload-identity helpers); providers delegate to this package.
- JWT signing: IamClient gained SignJwt; SignedJWTForVault / GenerateSignedJWTForVault construct JWTs with sub/aud/iat/exp and obtain signatures from GCP IAM.
- Vault provider: stores gcpSignedJWT, uses loginWithSignedJWT to authenticate to Vault with the signed JWT, and records gcpServiceAccountEmail where applicable.
- Secret Manager provider: delegates token sourcing to runtime/gcp/auth and removed prior in-file workload-identity logic.
- Tests & docs: Added unit tests under runtime/gcp/auth for token sources and JWT generation; removed older provider tests; added Vault GCP auth docs and example YAML.
- Modules: Large workspace go.mod/go.sum updates (multiple golang.org/x/*, google.* and OpenTelemetry bumps). A replace directive was used during development to reference the local GCP provider module, which caused many dependency file changes.

Notes
- Validated on GKE with Workload Identity and Vault GCP IAM auth; the change fixes the JWT parsing error.
- Review discussion flagged an architectural concern about provider-to-provider coupling; reviewer requested moving auth logic to runtime/gcp (auth package) — this PR implements that extraction, though some linter/Sonar feedback and cleanup were still outstanding in the thread.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Skarlso commented:
And the sonar issues? :D Please 🙇

SamuelMolling committed:
feat: refactor GCP auth to runtime and add Vault GCP Workload Identity support

This commit refactors shared GCP authentication logic into a new runtime/gcp/auth
module to eliminate co-dependencies between providers and adds support for GKE
Workload Identity authentication in Vault provider.

Changes:
- Create new runtime/gcp/auth module with shared authentication logic
- Extract NewTokenSource and Workload Identity functions from GCP Secret Manager
- Implement Vault GCP authentication using Workload Identity
- Reuse existing workloadIdentity.TokenSource() to ensure compatibility
- Update all providers to use the new shared runtime module
- Add proper module dependencies and replace directives

The Vault provider now supports GCP IAM authentication with Workload Identity,
using the same proven authentication flow as the GCP Secret Manager provider.

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
Merge branch 'main' into fix/vault-gcp-workload-identity-jwt

SamuelMolling committed:
docs: add GCP authentication method for Vault provider

Add documentation for GCP IAM authentication in Vault provider, supporting both
Workload Identity (recommended for GKE) and service account keys.

- Add vault-gcp-store.yaml snippet with example configuration
- Update hashicorp-vault.md to include gcpAuth in supported methods
- Add GCP authentication section following existing documentation patterns

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
chore: run go mod tidy to fix dependencies

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
docs: fix field name in vault-gcp-store.yaml comment

Use correct field name 'secretAccessKeySecretRef' instead of 'secretAccessKey'
to match the GCPSMAuthSecretRef struct.

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
chore: run go mod tidy in providers/v1/gcp

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
Merge branch 'main' of github.com:external-secrets/external-secrets into fix/vault-gcp-workload-identity-jwt

SamuelMolling committed:
chore: add runtime/gcp replace directive to generator modules

Add runtime/gcp replace directive to generators/v1/gcr and generators/v1/vault
to fix 'no go files to analyze' linting error. These generators depend on
providers that use runtime/gcp/auth, so they need the replace directive to
resolve the local module correctly.

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
chore: update golang.org/x/crypto to v0.46.0 to fix security vulnerabilities

Update golang.org/x/crypto from v0.43.0 to v0.46.0 to address:
- GHSA-j5w8-q4qc-rx2x: unbounded memory consumption in ssh
- GHSA-f6x5-jh6r-wrfv: panic if message is malformed due to out of bounds read

This update affects runtime/gcp and all dependent modules:
- providers/v1/gcp
- providers/v1/vault
- generators/v1/gcr
- generators/v1/vault

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
chore: propagate golang.org/x/* updates to root module

Run go mod tidy on root module to propagate security updates from
runtime/gcp, including golang.org/x/crypto v0.46.0 and related
golang.org/x/* dependencies.

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
chore: add runtime/gcp replace directive to e2e module

Add runtime/gcp replace directive to e2e/go.mod to fix 'unknown revision
runtime/gcp/v0.0.0' error when running 'make reviewable'. The e2e tests
import providers/v1/gcp which depends on runtime/gcp/auth.

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
refactor: rename runtime/gcp module to runtime/gcp/auth

- Renamed module from runtime/gcp to runtime/gcp/auth following code review suggestion
- Updated all go.mod files to reference runtime/gcp/auth instead of runtime/gcp
- Updated replace directives in all dependent modules
- Created go.sum for the new module path

Affected modules:
- go.mod (root)
- e2e/go.mod
- providers/v1/vault/go.mod
- providers/v1/gcp/go.mod
- generators/v1/vault/go.mod
- generators/v1/gcr/go.mod

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
chore: update go.mod/go.sum and remove old runtime/gcp module files

- Updated runtime/gcp/auth/go.mod and go.sum after go mod tidy
- Removed old runtime/gcp/go.mod and go.sum (replaced by runtime/gcp/auth)

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling commented on runtime/gcp/go.mod:
```diff
@@ -0,0 +1,101 @@
+module github.com/external-secrets/external-secrets/runtime/gcp
```
Still working well.

evrardjp commented on docs/snippets/vault-gcp-store.yaml:
```diff
@@ -0,0 +1,35 @@
+            serviceAccountRef:
+              name: "my-sa"
+              # Namespace is required for ClusterSecretStore
+              namespace: "default"
```
maybe silly question, can this be used to refer to a different namespace than the SecretStore? That looks alarming to me (Inconsistency between L29 and L5).  Is that on purpose?

evrardjp reviewed (commented):
I think I missed something: I see tests removed but not tests added.
If we consider this to be a stable interface, we need to have it properly tested. 

I am sorry I did not do a full review. I will do a full review when I have your answer about the needs. 

Side note: None of your commits messages are useful to me.
They all mention  _what you did_ (which you can guess by the title and looking at the content), not _why_ you did them. It makes it harder to review.

SamuelMolling committed:
test: add comprehensive unit tests for runtime/gcp/auth module

  Add unit tests with mocks for workload identity authentication:
  - Add auth_test.go with tests for token source creation and JWT generation
  - Add workload_identity_test.go with mocked IAM, metadata, and token generators
  - Add workload_identity_federation_test.go for federation configuration tests
  - Update go.mod with testify dependency
  - Add documentation links to workload_identity.go for Vault audience format
  - Fix vault-gcp-store.yaml to use ClusterSecretStore consistently

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
Merge remote-tracking branch 'origin/fix/vault-gcp-workload-identity-jwt' into fix/vault-gcp-workload-identity-jwt

Resolved conflict: kept deletion of workload_identity_federation_test.go
as the code was moved to runtime/gcp/auth/

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
refactor: remove unused NumericProjectIDWithContext method

Removed NumericProjectIDWithContext from MetadataClient interface
as it was never called anywhere in the codebase. Only
ProjectIDWithContext and InstanceAttributeValueWithContext are
actively used.

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
refactor: remove unused iamClient field from workloadIdentity struct

The iamClient field was always nil since IAM clients are created
on-demand and closed immediately after use. Removed the field and
simplified the Close() method to just return nil with a comment
explaining why.

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
refactor: remove unused clusterProjectID field from workloadIdentity struct

- Removed clusterProjectID field from workloadIdentity struct as it was
  never read by any method (gcpWorkloadIdentity reads from parameter)
- Removed projectID parameter from newWorkloadIdentity() constructor
- Kept projectID parameter in exported functions for API compatibility
- Added nolint directive for unused parameter warning

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
fix: address golangci-lint issues in runtime/gcp/auth

- Remove unused mockIamClient struct and methods from workload_identity_test.go
- Remove unused imports (credentialspb, gax)
- Replace lambda functions with direct function references (unlambda)
- Fix comment to end with period (godot)
- Break long function signature into multiple lines (golines)
- Rename unused projectID parameter to _ for API compatibility

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
test: add test case for GCP SA annotation path in TokenSource

Add test that exercises the ServiceAccount annotation path:
- Create ServiceAccount with 'iam.gke.io/gcp-service-account' annotation
- Add iamClientCreator field to workloadIdentity struct for dependency injection
- Inject mockIamClient implementing GenerateAccessToken
- Assert non-error token source is returned

This enables testing the on-demand IAM client creation and GenerateAccessToken
flow when the GCP SA annotation is present on the Kubernetes ServiceAccount.

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
fix: make tests independent of K8s cluster

Refactor saTokenGenerator creation to use a factory function that can be
mocked in tests. This allows all tests in runtime/gcp/auth to run without
requiring a real Kubernetes cluster or KUBERNETES_MASTER environment variable.

Changes:
- Add newSATokenGeneratorFunc variable that can be overridden in tests
- Add functional options to newWorkloadIdentity and newWorkloadIdentityFederation
- Create testing_helpers_test.go with TestMain to set up mock factory
- Update all tests to use mock saTokenGenerator

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
test: address code review feedback

- Add positive test case for SignedJWTForVault that exercises the happy path
  with mock IAM client injection to verify JWT signing logic
- Remove duplicate mockWifSATokenGenerator and reuse testSATokenGenerator
  from testing_helpers_test.go to reduce code duplication
- Verify that SignJwt is called with correct service account and audience format

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
refactor: address nitpick comments from code review

- Remove unused validateCalled field from TestSignedJWTForVault test struct
- Add explanatory comment to TestWorkloadIdentityClose clarifying that
  IAM clients are created on-demand and closed immediately, so Close()
  exists only for interface compatibility

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

Skarlso commented on providers/v1/gcp/secretmanager/auth.go:
```diff
@@ -23,64 +23,22 @@ package secretmanager
-	return config.TokenSource(ctx), nil
+// GenerateSignedJWTForVault generates a signed JWT specifically for Vault GCP IAM authentication.
+// This is a wrapper around the shared runtime/gcp/auth implementation.
+func GenerateSignedJWTForVault(ctx context.Context, wi *esv1.GCPWorkloadIdentity, role, projectID, storeKind string, kube kclient.Client, namespace string) (string, error) {
```
This function is not used. 🤔

Skarlso commented on runtime/gcp/auth/auth.go:
```diff
@@ -0,0 +1,128 @@
+func GenerateSignedJWTForVault(
+	ctx context.Context,
+	wi *esv1.GCPWorkloadIdentity,
+	role, _ /* projectID */, storeKind string,
```
Why do you have _projectID_ if it's not used? Just remove it. It's not an interface as far as I can see.

Skarlso reviewed (commented):
Still a few questions about the approach and code. :)

SamuelMolling committed:
Update providers/v1/vault/auth_gcp.go

Co-authored-by: Gergely Bräutigam <skarlso777@gmail.com>
Signed-off-by: SamuelMolling <56612163+SamuelMolling@users.noreply.github.com>

SamuelMolling committed:
refactor: improve GCP workload identity code quality

- Remove unused GenerateSignedJWTForVault wrapper from secretmanager/auth.go
- Remove unused projectID parameter from GenerateSignedJWTForVault
- Implement refreshableTokenSource for automatic token refresh
- Extract getIdentityBindingToken to eliminate duplicate IAM client creation
- Improve error messages with more context (path, role, gcpServiceAccount)
- Address code review feedback for cleaner architecture

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
fix: add mutex for thread safety in refreshableTokenSource

The Token() method reads and writes currentToken which can be accessed
concurrently if the token source is shared across goroutines. Added
sync.Mutex to ensure thread-safe access.

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
refactor: address nitpick comments for workload identity

- Document context storage limitation in refreshableTokenSource
- Add 'iat' (issued at) claim to JWT payload for better debugging
- Improve idbindtoken error message to include response body

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
refactor: improve observability and consistency in workload identity

- Extract jwtTTL constant (15 minutes) for SA tokens and Vault JWTs
- Add CallGCPSMSignJwt constant for metrics
- Add metrics observation for SignJwt API call
- Improve error message for idbindtoken to include response body

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
docs: improve code comments for maintainability

- Document why Close() errors are ignored in IAM client defers
- Add detailed comments in TestMain explaining cleanup pattern
- Include example of t.Cleanup for individual test customization

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
test: add mock for ADC in unit tests

- Add defaultTokenSourceFunc variable for dependency injection
- Create mockTokenSource for testing without real GCP credentials
- Update TestNewTokenSource and TestTokenSourceFallback to use mocks
- Ensure tests pass in CI environments without ADC configured

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

jakobmoellerdev commented on runtime/gcp/auth/workload_identity.go:
```diff
@@ -115,22 +123,87 @@ type saTokenGenerator interface {
+// the token when it's about to expire. This ensures long-running operations
+// don't fail due to token expiration.
+//
+// Note: The ctx field is stored because oauth2.TokenSource.Token() doesn't accept
```
I dont understand this. why cant you just have a background context that has a bounded timeout or dedicated shutdown? it feels like reusing the context here is artificially introduced as a limit.

jakobmoellerdev reviewed (commented):
generally okay for me but shared modules need a little better descriptions and guidance. i.e. a readme or developer guide section.

SamuelMolling committed:
refactor: use bounded context for token refresh operations

- Remove stored context from refreshableTokenSource struct
- Use context.Background() with 30s timeout for each refresh operation
- This ensures token refresh works independently of original request context
- Addresses review feedback about artificial context limitation

Signed-off-by: Samuel Molling <samuelmolling@gmail.com>

SamuelMolling committed:
Merge branch 'main' into fix/vault-gcp-workload-identity-jwt

Skarlso committed:
fix: remove the submodule changes and just use the runtime package

Signed-off-by: Gergely Brautigam <182850+Skarlso@users.noreply.github.com>

Skarlso committed:
remove all the go mod changes and let's see what breaks

Signed-off-by: Gergely Brautigam <182850+Skarlso@users.noreply.github.com>

Skarlso committed:
running go mod tidy here and there

Signed-off-by: Gergely Brautigam <182850+Skarlso@users.noreply.github.com>

Skarlso committed:
run go mod tidy everywhere

Signed-off-by: Gergely Brautigam <182850+Skarlso@users.noreply.github.com>

Skarlso commented:
/ok-to-test sha=dcea3eced3c644c6f72810bc324ab234c0589c66

Skarlso commented:
Depds update needs to happen because of the runtime update I'm afraid. Because of GCP we have new dependencies, that will bleed out into the other modules because of the `../runtime` rewrites everywhere.
